neutron_qos/services/qos/common/htb.py

```diff
@@ -0,0 +1,462 @@
+            elif off == 12:
+                address = socket.inet_ntoa(data[p+4:p+8])
+                prefix = MASK2PREFIXLEN_MAP[mask]
+                qos_filter['src_addr'] = '%s/%s' % (address, prefix)
+            elif off == 16:
+                address = socket.inet_ntoa(data[p+4:p+8])
+                prefix = MASK2PREFIXLEN_MAP[mask]
+                qos_filter['dst_addr'] = '%s/%s' % (address, prefix)
+        return qos_filter
+
+    def parse_msg(self, data):
+        super(RecvTFilterMsg, self).parse_msg(data)
+        name = self.attr_data_to_name(TCA_ACT_KIND)
+        if name == 'u32':
+            if TCA_ACT_OPTIONS not in self.attrs:
+                self.has_u32_sel = False
+            else:
+                nlattrs = self.parse_nlattrs(self.attrs[TCA_ACT_OPTIONS])
+                if TCA_U32_SEL in nlattrs and TCA_U32_CLASSID in nlattrs:
+                    self.has_u32_sel = True
+                    self.classid = self.template._handle_to_str(
+                        struct.unpack('I', nlattrs[TCA_U32_CLASSID])[0])
+                    qos_filter = self.data_to_qos_filter(nlattrs[TCA_U32_SEL])
+                    qos_filter['prio'] = self.template.prio
+                    self.qos_filter_dict = qos_filter
+                else:
+                    self.has_u32_sel = False
+
+
+class RTNetLink(netns.NetNSSwitcher):
+    _recv_msg_class = {
+        RecvIfInfoMsg._msg_type: RecvIfInfoMsg,
+        RecvQdiscMsg._msg_type: RecvQdiscMsg,
+        RecvTClassMsg._msg_type: RecvTClassMsg,
+        RecvTFilterMsg._msg_type: RecvTFilterMsg,
+    }
+
+    def __init__(self, netns=None):
+        super(RTNetLink, self).__init__(netns)
+
+    def __enter__(self):
+        super(RTNetLink, self).__enter__()
+        self._sock = socket.socket(
+            socket.AF_NETLINK, socket.SOCK_RAW, socket.NETLINK_ROUTE)
+        self._sock.bind((self._pid, 0))
+        self._seq = int(time.time())
+        self._sfd = os.fdopen(self._sock.fileno(), 'w+b')
+        return self
+
+    def __exit__(self, e_type, e_value, traceback):
+        super(RTNetLink, self).__exit__(e_type, e_value, traceback)
+        self._sfd.close()
+        self._sock.close()
+
+    def send_and_recv(self, msg):
+        msgs = []
+        self._seq += 1
+        msg.header.seq = self._seq
+        self._sfd.write(msg.data)
+        recv_header = NLM_Header.from_data(self._sfd.read(NLM_Header.length))
+        if recv_header.msg_type != NLMSG_ERROR:
+            while recv_header.msg_type != NLMSG_DONE:
+                body_length = recv_header.length - NLM_Header.length
+                recv_msg = self._recv_msg_class[recv_header.msg_type](
+                    recv_header)
+                recv_msg.parse_msg(self._sfd.read(body_length))
+                msgs.append(recv_msg)
+                recv_header = NLM_Header.from_data(
+                    self._sfd.read(NLM_Header.length))
+        self._sfd.read(recv_header.length-NLM_Header.length)
```

Comment:
所有review done

2016-08-22 15:35 GMT+08:00 huntxu <notifications@github.com>:

> In neutron_qos/services/qos/common/htb.py
> <https://github.com/eayunstack/neutron-qos/pull/17#discussion_r75629383>:
>
> > +    def send_and_recv(self, msg):
> > +        msgs = []
> > +        self._seq += 1
> > +        msg.header.seq = self._seq
> > +        self._sfd.write(msg.data)
> > +        recv_header = NLM_Header.from_data(self._sfd.read(NLM_Header.length))
> > +        if recv_header.msg_type != NLMSG_ERROR:
> > +            while recv_header.msg_type != NLMSG_DONE:
> > +                body_length = recv_header.length - NLM_Header.length
> > +                recv_msg = self._recv_msg_class[recv_header.msg_type](
> > +                    recv_header)
> > +                recv_msg.parse_msg(self._sfd.read(body_length))
> > +                msgs.append(recv_msg)
> > +                recv_header = NLM_Header.from_data(
> > +                    self._sfd.read(NLM_Header.length))
> > +        self._sfd.read(recv_header.length-NLM_Header.length)
>
> 出错和正常最后的一part信息还在fd里面，此时最后一部分信息的头已经读完，把最后这部分消息的消息体也给读掉，这是完整地读取过程的最后一步啊，
> 当然也算是清空buffer吧。
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/eayunstack/neutron-qos/pull/17/files/b6a33584ad195955de06c66df75d0087cd592933..f317640e7759cf79a60f68f8896ca756a966b27f#r75629383>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGUT0Tcxel7Tg0I3Xpe4tK6gqR54Vsu3ks5qiVE1gaJpZM4JW3dU>
> .
>